CompareRaftMembers/DiffStream.js

```diff
@@ -0,0 +1,173 @@
+                // buffer gets too big: for this we may also need to
+                // split the input buffer at the last bucket buffer's
+                // key, so that the comparison is meaningful
+                if (bucketBuffer.length === this.maxBufferSize) {
+                    const maxKey = bucketBuffer[bucketBuffer.length - 1].key;
+                    let inputBufferLimit = this.inputBuffer.findIndex(item => item.fullKey > maxKey);
+                    if (inputBufferLimit === -1) {
+                        inputBufferLimit = this.inputBuffer.length;
+                    }
+                    const inputBuffer = this.inputBuffer.splice(0, inputBufferLimit);
+                    this._compareBuffers(inputBuffer, bucketBuffer);
+                    bucketBuffer = [];
+                }
+            })
+            .on('end', () => {
+                const { inputBuffer } = this;
+                this.inputBuffer = [];
+                this._compareBuffers(inputBuffer, bucketBuffer);
+                callback();
+            })
+            .on('error', err => {
+                // unrecoverable error after retries: raise the error
+                this.emit('error', err);
+            });
+    }
+
+    _processItem(itemInfo, callback) {
+        const { bucketName } = itemInfo;
+
+        const isNewBucket = (this.inputBufferBucket && bucketName !== this.inputBufferBucket);
+        if (isNewBucket || this.inputBuffer.length === this.maxBufferSize) {
+            const lastObjectKey = isNewBucket
+                ? null
+                : this.inputBuffer[this.inputBuffer.length - 1].objectKey;
+            this._compareInputBufferWithBucketd(lastObjectKey, () => {
+                this.inputBufferBucket = bucketName;
+                this.currentMarker = lastObjectKey;
+                this._ingestItem(itemInfo, callback);
+            });
+        } else {
+            this.inputBufferBucket = bucketName;
+            // due to the workaround described above, we may be
+            // processing a new bucket's key here and need to reset
+            // the marker in this case only, without querying bucketd
+            if (isNewBucket) {
+                this.currentMarker = null;
+            }
+            this._ingestItem(itemInfo, callback);
+        }
+    }
+
+    _ingestItem(itemInfo, callback) {
+        const { fullKey } = itemInfo;
+        this.inputBuffer.push(itemInfo);
+        return callback();
+    }
+
+    _transform(item, encoding, callback) {
+        const itemInfo = this._parseItem(item);
+        this._processItem(itemInfo, callback);
+    }
+
+    _flush(callback) {
+        if (this.inputBuffer.length > 0) {
+            this._compareInputBufferWithBucketd(null, callback);
+        } else {
+            callback();
+        }
+    }
+
+    _parseItem(item) {
+        const { key: fullKey, value } = item;
+        const slashIndex = fullKey.indexOf('/');
```

Comment:
want to make sure that bucketName can't be prefixed with '-' for some reason as that would mess up this split.

Reply:
yes, bucket names are DNS-compatible, hence no slash or other forbidden character